### PR TITLE
I 128 mobile app

### LINKED ISSUE #128: Create mobile app
1. Blueprint for mobile app.
2. Successful app user registration with mobile/email verification based on user preference.
3. Successful login into the app using username and password.

### PR DESCRIPTION
**Pull request by EbrahimAli-RU**
Fixes #128 

**Implement User Authentication and Initial App Setup**
- Set up the project using Expo, laying the groundwork for future development.
- Developed the login screen with AWS Cognito integration for user authentication.
- Added user registration functionality and OTP verification process.
- Integrated AWS Cognito for handling signup and OTP submission.
- Implemented navigation flow from registration to OTP verification screens.

- [x] Does not depend on other PRs
- [ ] Depends on other PR being merged first: #

## What kind of change does this PR introduce (check at least one)

- [ ] Bugfix
- [ ] Improvement
- [x] New feature
- [ ] Refactor
- [ ] Infrastructure
- [ ] Other (please specify)

## How would you describe the changes in a changelog (1-2 lines)?

Provide a brief overview of the changes introduced in this pull request.

## Related Issues/PRs (optional)

Link to the relevant issue(s) or provide a clear description of the problem or feature being addressed.

## How to test

- [ ] Needs regression testing (add QA needed label)

## Checklist

- [x] I have tested these changes locally and they function as expected.
- [ ] I have added or updated necessary documentation to reflect the changes.
- [ ] I have followed the project's coding style and guidelines.
- [ ] I have added appropriate test coverage to ensure the changes are functioning correctly.
- [ ] I have checked for any potential conflicts with other branches or pull requests.
- [ ] I have updated necessary OpenAPI specs

## Reviewer Request (optional)

If you have any specific requests or suggestions for reviewers, mention them here.

## Screenshots (optional)

If applicable, provide any additional screenshots or GIFs that showcase the changes made in this pull request.

## Demo Link (optional)

If applicable, provide a link to a live demo or deployment where the changes can be reviewed.

## Known Issues (optional)

If there are any known issues or limitations with the changes made in this pull request, describe them here.

## Additional Notes (optional)

Add any additional notes or comments you think are relevant to this pull request.

## Deploy risk & follow up

Please assess risk & follow up on deployment of this PR. Elaborate your choice if selecting anything other than 'No risk'.

- [x] No risk
- [ ] Low
- [ ] High

### CODE REVIEW

File: clients/mobile/src/utility/validator.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  return !emailRegex.test(value) ? 'Invalid email address' : ''
+}
+
+export const isValidPhoneNumber = (value: string): string => {
+  const phoneRegex = /^\+?[1-9]\d{1,14}$/
+  return !phoneRegex.test(value) ? 'Invalid phone number' : ''
+}
+
+export const isValidPassword = (value: string): string => {
+  let errorMsg = ''
```
[the remark this answers is not in the document]
Comment (sakibstark11) [reply clients/mobile/src/utility/validator.ts:32]: I'll add examples in a bit

[the remark this answers is not in the document]
Comment (sakibstark11) [reply clients/mobile/src/utility/validator.ts:32]: I don't really understand this. You're creating a giant string, how are we going to show this on the ui.
Shouldn't it be split into multiple lines?

File: clients/mobile/src/utility/validator.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  const emailRegex = /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i
+  return !emailRegex.test(value) ? 'Invalid email address' : ''
+}
+
+export const isValidPhoneNumber = (value: string): string => {
+  const phoneRegex = /^\+?[1-9]\d{1,14}$/
+  return !phoneRegex.test(value) ? 'Invalid phone number' : ''
+}
+
+export const isValidPassword = (value: string): string => {
```
Comment (sakibstark11) [inline clients/mobile/src/utility/validator.ts:31]: the name of the function suggests that if the password is valid, we would get a truthy value returned, if not we might get errors array. But in your case, I struggle to follow what it's doing.

File: clients/mobile/src/utility/validator.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (sakibstark11) [reply clients/mobile/src/utility/validator.ts:32]:
```ts
const isValidPassword = (value) => {
  const tests = [
    {
      test: value.length >= passwordPolicy.minimum_length,
      error: `Min ${passwordPolicy.minimum_length} chars`,
    },
    {
      test: !passwordPolicy.require_lowercase || /[a-z]/.test(value),
      error: "1 lowercase",
    },
    {
      test: !passwordPolicy.require_uppercase || /[A-Z]/.test(value),
      error: "1 uppercase",
    },
    {
      test: !passwordPolicy.require_numbers || /\d/.test(value),
      error: "1 number",
    },
    {
      test:
        !passwordPolicy.require_symbols || /[!@#$%^&*(),.?":{}|<>]/.test(value),
      error: "1 symbol",
    },
  ]

  const errors = tests
    .filter(({ test }) => !test)
    .map(({ error }) => error)
    .join(", ")

  return errors ? `Password must contain: ${errors}` : true
}
```
I am thinking something like this. Infact I would return the error array and let the viewModel handle how they want to do it.

[the remark this answers is not in the document]
Comment (sakibstark11) [reply clients/mobile/src/utility/validator.ts:32]: actually the passwordPolicy should contain how it wants to test it

File: clients/mobile/src/utility/validator.ts (hunk 2 above)
Comment (EbrahimAli-RU) [reply clients/mobile/src/utility/validator.ts]: Update the name of the function.

You will find with this [Commit](https://github.com/CraftsmenLtd/BloodConnect/pull/138/commits/94e59fc445b4f8c16e12ee97341d0810f25e56ab)

File: clients/mobile/src/utility/validator.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (EbrahimAli-RU) [reply clients/mobile/src/utility/validator.ts]: I think the decision is still going on how we will show the password field error. Will it be okay to change now or should I keep things as it is.

You will find with this [Commit](https://github.com/CraftsmenLtd/BloodConnect/pull/138/commits/94e59fc445b4f8c16e12ee97341d0810f25e56ab)

File: clients/mobile/src/utility/validator.ts [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  const emailRegex = /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i
+  return !emailRegex.test(value) ? 'Invalid email address' : ''
+}
+
+export const validatePhoneNumberAndGetErrorMessage = (value: string): string => {
+  const phoneRegex = /^\+?[1-9]\d{1,14}$/
+  return !phoneRegex.test(value) ? 'Invalid phone number' : ''
+}
+
+export const validatePasswordAndGetErrorMessage = (value: string): string => {
```
Comment (sakibstark11) [inline clients/mobile/src/utility/validator.ts:31]: maybe this should be called checkErrorsInPassword and it should return null if there are no errors

File: clients/mobile/src/utility/validator.ts [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    .filter(({ test }) => !test)
+    .map(({ error }) => error)
+    .join(', ')
+
+  return (errors !== '') ? `Password must contain: ${errors}` : ''
+}
+
+export type ValidationRule = (value: string) => string
+
+export const validateInput = (value: string, rules: ValidationRule[]): string => {
```
Comment (sakibstark11) [inline clients/mobile/src/utility/validator.ts:65]: same here

File: clients/mobile/src/utility/validator.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  for (const rule of rules) {
```
Comment (sakibstark11) [inline clients/mobile/src/utility/validator.ts:66]: can just use forEach, this is 2024

File: clients/mobile/src/setup/theme/context/ThemeContext.tsx [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+
+export const ThemeContext = createContext<Theme | null>(null)
+
+interface ThemeProviderProps {
+  children: ReactNode;
+}
+
+export const ThemeProvider: React.FC<ThemeProviderProps> = ({ children }) => {
+  const systemTheme = useColorScheme()
+  const [theme, setTheme] = useState<Theme>(systemTheme === 'dark' ? darkTheme : lightTheme)
```
Comment (sakibstark11) [inline clients/mobile/src/setup/theme/context/ThemeContext.tsx:13]: this should be const, the theme name

File: clients/mobile/src/utility/validator.ts (hunk 5 above)
Comment (EbrahimAli-RU) [reply clients/mobile/src/utility/validator.ts:66]: I want to return from this function if I get an error. I don't want to iterate over the entire `rules` array even after getting an error. If I use `forEach`, I think it's not possible to achieve that.

File: clients/mobile/src/setup/theme/context/ThemeContext.tsx (hunk 6 above)
Comment (EbrahimAli-RU) [reply clients/mobile/src/setup/theme/context/ThemeContext.tsx:13]: Don't understand this vai should I change this to const or theme to `themeName`. But as I can see this is already const.

File: clients/mobile/src/utility/validator.ts (hunk 3 above)
Comment (EbrahimAli-RU) [reply clients/mobile/src/utility/validator.ts]: Update the name to `checkErrorsInPassword ` and return null if the error is not found.[Commit](https://github.com/CraftsmenLtd/BloodConnect/pull/138/commits/0e9819bab559227cbd4d26afdb38b6451c5b83ff)

File: clients/mobile/src/utility/validator.ts (hunk 4 above)
Comment (EbrahimAli-RU) [reply clients/mobile/src/utility/validator.ts]: Can you please suggest a name I can't think better name?

File: clients/mobile/src/setup/theme/context/ThemeContext.tsx (hunk 6 above)
Comment (sakibstark11) [reply clients/mobile/src/setup/theme/context/ThemeContext.tsx:13]: The keyword dark should be a constant

File: clients/mobile/src/utility/validator.ts (hunk 5 above)
Comment (sakibstark11) [reply clients/mobile/src/utility/validator.ts:66]: I see, that's a valid argument

**Review by melon-ruet — CHANGES_REQUESTED** [review]

**Review by sakibstark11 — APPROVED** [review]

**Review by sakibstark11 — APPROVED** [review]

File: clients/mobile/.gitignore
[the remark this answers is not in the document]
Comment (EbrahimAli-RU) [reply clients/mobile/.gitignore:1]: This was created by default when initiating the mobile app that's why I didn't remove that.

**Review by KhirulIslam — APPROVED** [review]

**Pull request merged**